### PR TITLE
Attemp to resolved cream script

### PR DESCRIPTION
**Pull request by petrovska-petro**
Updates on the cream script for accounting deposited collateral in: 
- Iron Bank 
- Ethereum
- BSC
- Fantom

Also attempt to grab all the staking services from: 
- BSC 
- ETH2 balances thru accumulated() method in CRETH2 + pendant balance of ETH on the CRETH2 contract
- Fantom (note: not sure on the error for the graphQLClient.request() method, but I believe that calculation it the one needed to grab the total amount staked * current_fantom_price_usd)

### CODE REVIEW

File: projects/cream.js [hunk 1]
```diff
@@ -1,36 +1,158 @@
...
+  );
+
   await Promise.all(
-    tokens.data.map(async token => {
-      let balance = (token.underlying_price.value * token.exchange_rate.value * token.total_supply.value * price_feed.data.ethereum.usd);
+    tokens_ethereum.data.map(async (token) => {
+      let balance =
+        token.underlying_price.value *
+        token.exchange_rate.value *
+        token.total_supply.value *
```
Comment (jeremyHD) [inline projects/cream.js:38]:
```suggestion
        token.cash.value *
```

File: projects/cream.js [hunk 2]
```diff
@@ -1,36 +1,158 @@
...
+        price_feed.data.ethereum.usd;
+
+      tvl += balance;
+    }),
+    tokens_bsc.data.map(async (token) => {
+      let balance =
+        token.underlying_price.value *
+        token.exchange_rate.value *
+        token.total_supply.value *
```
Comment (jeremyHD) [inline projects/cream.js:47]:
```suggestion
        token.cash.value *
```

File: projects/cream.js [hunk 3]
```diff
@@ -1,36 +1,158 @@
...
+        price_feed.data.binancecoin.usd;
+      tvl += balance;
+    }),
+    tokens_fantom.data.map(async (token) => {
+      let balance =
+        token.total_supply.value *
+        token.underlying_price.value *
+        token.exchange_rate.value;
```
Comment (jeremyHD) [inline projects/cream.js:55]:
```suggestion
        token.cash.value;
```

File: projects/cream.js [hunk 4]
```diff
@@ -1,36 +1,158 @@
...
+      tvl += balance;
+    }),
+    tokens_iron_bank.data.map(async (token) => {
+      let balance =
+        token.underlying_price.value *
+        token.exchange_rate.value *
+        token.total_supply.value;
```
Comment (jeremyHD) [inline projects/cream.js:62]:
```suggestion
        token.cash.value;
```

**Comment by jeremyHD** [comment]
Hey @petrovska-petro - thanks for helping CREAM on fixing the TVL script :)
Have left some comment on review. Please take a look!

**Comment by petrovska-petro** [comment]
Hi @jeremyHD, I will change those lines in the next commit and add also the right decimals for each token.

@0xngmi it seems that many of the current usd prices are not available in @defillama/sdk currently, I think it should be adapted to have the right usd values for all the tokens present in the lending markets, iron bank and similar in Cream Finance to be able to have an accurate tvl over time.  Otherwise, they will be set to 0 as default I believe.

I encountered that sometimes for some reason the getCash() method from the CErc20's abi seems to misreport the value asked, which could lead to miscalculation later on on the tvl for a specific token. @jeremyHD could that be possible?.

It happened in the following (cyDAI: "0x8e595470Ed749b85C6F7669de83EAe304C2ec68F"), it was reporting more than held in the contract at least when I checked via @defillama/sdk.

**Comment by 0xngmi** [comment]
> @0xngmi it seems that many of the current usd prices are not available in @defillama/sdk currently, I think it should be adapted to have the right usd values for all the tokens present in the lending markets, iron bank and similar in Cream Finance to be able to have an accurate tvl over time. Otherwise, they will be set to 0 as default I believe.

Which tokens are you seeing that on? If it's on cyTokens you might need to unwrap them to get their underlying

**Comment by jeremyHD** [comment]
> I encountered that sometimes for some reason the getCash() method from the CErc20's abi seems to misreport the value asked, which could lead to miscalculation later on on the tvl for a specific token. @jeremyHD could that be possible?.
> 
> It happened in the following (cyDAI: "0x8e595470Ed749b85C6F7669de83EAe304C2ec68F"), it was reporting more than held in the contract at least when I checked via @defillama/sdk.

Sorry I'm not familiar with @defillama/sdk, but I've just look at cyDAI's cash on both etherscan and our own API it does match.
Would you mind to provide more details? Like how do you call the sdk? Thanks

**Comment by petrovska-petro** [comment]
@0xngmi Here is the list from the output which defaults to 0 while running. I believe some major portion of it are already in underlying denomination

``Couldn't find the price of token at USDT, assuming a price of 0 for it...
Couldn't find the price of token at USDC, assuming a price of 0 for it...
Couldn't find the price of token at YFI, assuming a price of 0 for it...
Couldn't find the price of token at BAL, assuming a price of 0 for it...
Couldn't find the price of token at COMP, assuming a price of 0 for it...
Couldn't find the price of token at YCRV, assuming a price of 0 for it...
Couldn't find the price of token at LEND, assuming a price of 0 for it...
Couldn't find the price of token at CRV, assuming a price of 0 for it...
Couldn't find the price of token at MTA, assuming a price of 0 for it...
Couldn't find the price of token at YYCRV, assuming a price of 0 for it...
Couldn't find the price of token at FTT, assuming a price of 0 for it...
Couldn't find the price of token at YETH, assuming a price of 0 for it...
Couldn't find the price of token at SRM, assuming a price of 0 for it...
Couldn't find the price of token at UNI, assuming a price of 0 for it...
Couldn't find the price of token at WNXM, assuming a price of 0 for it...
Couldn't find the price of token at SWAG, assuming a price of 0 for it...
Couldn't find the price of token at CEL, assuming a price of 0 for it...
Couldn't find the price of token at DPI, assuming a price of 0 for it...
Couldn't find the price of token at BBTC, assuming a price of 0 for it...
Couldn't find the price of token at BOND, assuming a price of 0 for it...
Couldn't find the price of token at KP3R, assuming a price of 0 for it...
Couldn't find the price of token at HBTC, assuming a price of 0 for it...
Couldn't find the price of token at HFIL, assuming a price of 0 for it...
Couldn't find the price of token at CRETH2, assuming a price of 0 for it...
Couldn't find the price of token at ESD, assuming a price of 0 for it...
Couldn't find the price of token at COVER, assuming a price of 0 for it...
Couldn't find the price of token at OMG, assuming a price of 0 for it...
Couldn't find the price of token at UNIV2WBTCETH, assuming a price of 0 for it...
Couldn't find the price of token at UNIV2ETHUSDT, assuming a price of 0 for it...
Couldn't find the price of token at UNIV2USDCETH, assuming a price of 0 for it...
Couldn't find the price of token at UNIV2DAIETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPWBTCETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPDAIETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPUSDCETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPETHUSDT, assuming a price of 0 for it...
Couldn't find the price of token at SLPSUSHIETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPYFIETH, assuming a price of 0 for it...
Couldn't find the price of token at WBTC, assuming a price of 0 for it...
Couldn't find the price of token at SNX, assuming a price of 0 for it...
Couldn't find the price of token at SUSD, assuming a price of 0 for it...
Couldn't find the price of token at PICKLE, assuming a price of 0 for it...
Couldn't find the price of token at AKRO, assuming a price of 0 for it...
Couldn't find the price of token at BBADGER, assuming a price of 0 for it...
Couldn't find the price of token at OGN, assuming a price of 0 for it...
Couldn't find the price of token at AMP, assuming a price of 0 for it...
Couldn't find the price of token at UST, assuming a price of 0 for it...
Couldn't find the price of token at FTM, assuming a price of 0 for it...
Couldn't find the price of token at PERP, assuming a price of 0 for it...
Couldn't find the price of token at YVECRV, assuming a price of 0 for it...
Couldn't find the price of token at YVSTECRV, assuming a price of 0 for it...
Couldn't find the price of token at YVCurveIB, assuming a price of 0 for it...
Couldn't find the price of token at RARI, assuming a price of 0 for it...
Couldn't find the price of token at SFI, assuming a price of 0 for it...
Couldn't find the price of token at YVCurvesETH, assuming a price of 0 for it...
Couldn't find the price of token at YVCurvestETH, assuming a price of 0 for it...
Couldn't find the price of token at ARNXM, assuming a price of 0 for it...
Couldn't find the price of token at MLN, assuming a price of 0 for it...
Couldn't find the price of token at VSP, assuming a price of 0 for it...
Couldn't find the price of token at G, assuming a price of 0 for it...
Couldn't find the price of token at Y3CRV, assuming a price of 0 for it...
Couldn't find the price of token at DUSD, assuming a price of 0 for it...
Couldn't find the price of token at EURS, assuming a price of 0 for it...
Couldn't find the price of token at GUSD, assuming a price of 0 for it...
Couldn't find the price of token at CUSDT, assuming a price of 0 for it...
Couldn't find the price of token at CUSDC, assuming a price of 0 for it...``

**Comment by petrovska-petro** [comment]
> > I encountered that sometimes for some reason the getCash() method from the CErc20's abi seems to misreport the value asked, which could lead to miscalculation later on on the tvl for a specific token. @jeremyHD could that be possible?.
> > It happened in the following (cyDAI: "0x8e595470Ed749b85C6F7669de83EAe304C2ec68F"), it was reporting more than held in the contract at least when I checked via @defillama/sdk.
> 
> Sorry I'm not familiar with @defillama/sdk, but I've just look at cyDAI's cash on both etherscan (DAI balance and getCash() ) and our own API it does match.
> Would you mind to provide more details? Like how do you call the sdk? Thanks

@jeremyHD all good now, getting the right values, it was a bug from my side on the invocation on things.

In case you'd like to run it from the console run the following: npx @defillama/sdk projects/creamswap/index.js

**Comment by petrovska-petro** [comment]
> > @0xngmi it seems that many of the current usd prices are not available in @defillama/sdk currently, I think it should be adapted to have the right usd values for all the tokens present in the lending markets, iron bank and similar in Cream Finance to be able to have an accurate tvl over time. Otherwise, they will be set to 0 as default I believe.
> 
> Which tokens are you seeing that on? If it's on cyTokens you might need to unwrap them to get their underlying

@0xngmi Here is the list from the output which defaults to 0 while running. I believe some major portion of it are already in underlying denomination

``Couldn't find the price of token at ETH, assuming a price of 0 for it...
Couldn't find the price of token at USDT, assuming a price of 0 for it...
Couldn't find the price of token at USDC, assuming a price of 0 for it...
Couldn't find the price of token at YFI, assuming a price of 0 for it...
Couldn't find the price of token at BAL, assuming a price of 0 for it...
Couldn't find the price of token at COMP, assuming a price of 0 for it...
Couldn't find the price of token at YCRV, assuming a price of 0 for it...
Couldn't find the price of token at LEND, assuming a price of 0 for it...
Couldn't find the price of token at CRV, assuming a price of 0 for it...
Couldn't find the price of token at MTA, assuming a price of 0 for it...
Couldn't find the price of token at YYCRV, assuming a price of 0 for it...
Couldn't find the price of token at FTT, assuming a price of 0 for it...
Couldn't find the price of token at YETH, assuming a price of 0 for it...
Couldn't find the price of token at SRM, assuming a price of 0 for it...
Couldn't find the price of token at UNI, assuming a price of 0 for it...
Couldn't find the price of token at WNXM, assuming a price of 0 for it...
Couldn't find the price of token at SWAG, assuming a price of 0 for it...
Couldn't find the price of token at CEL, assuming a price of 0 for it...
Couldn't find the price of token at DPI, assuming a price of 0 for it...
Couldn't find the price of token at BBTC, assuming a price of 0 for it...
Couldn't find the price of token at BOND, assuming a price of 0 for it...
Couldn't find the price of token at KP3R, assuming a price of 0 for it...
Couldn't find the price of token at HBTC, assuming a price of 0 for it...
Couldn't find the price of token at HFIL, assuming a price of 0 for it...
Couldn't find the price of token at CRETH2, assuming a price of 0 for it...
Couldn't find the price of token at ESD, assuming a price of 0 for it...
Couldn't find the price of token at COVER, assuming a price of 0 for it...
Couldn't find the price of token at OMG, assuming a price of 0 for it...
Couldn't find the price of token at UNIV2WBTCETH, assuming a price of 0 for it...
Couldn't find the price of token at UNIV2ETHUSDT, assuming a price of 0 for it...
Couldn't find the price of token at UNIV2USDCETH, assuming a price of 0 for it...
Couldn't find the price of token at UNIV2DAIETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPWBTCETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPDAIETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPUSDCETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPETHUSDT, assuming a price of 0 for it...
Couldn't find the price of token at SLPSUSHIETH, assuming a price of 0 for it...
Couldn't find the price of token at SLPYFIETH, assuming a price of 0 for it...
Couldn't find the price of token at WBTC, assuming a price of 0 for it...
Couldn't find the price of token at SNX, assuming a price of 0 for it...
Couldn't find the price of token at SUSD, assuming a price of 0 for it...
Couldn't find the price of token at PICKLE, assuming a price of 0 for it...
Couldn't find the price of token at AKRO, assuming a price of 0 for it...
Couldn't find the price of token at BBADGER, assuming a price of 0 for it...
Couldn't find the price of token at OGN, assuming a price of 0 for it...
Couldn't find the price of token at AMP, assuming a price of 0 for it...
Couldn't find the price of token at UST, assuming a price of 0 for it...
Couldn't find the price of token at FTM, assuming a price of 0 for it...
Couldn't find the price of token at PERP, assuming a price of 0 for it...
Couldn't find the price of token at YVECRV, assuming a price of 0 for it...
Couldn't find the price of token at YVSTECRV, assuming a price of 0 for it...
Couldn't find the price of token at YVCurveIB, assuming a price of 0 for it...
Couldn't find the price of token at RARI, assuming a price of 0 for it...
Couldn't find the price of token at SFI, assuming a price of 0 for it...
Couldn't find the price of token at YVCurvesETH, assuming a price of 0 for it...
Couldn't find the price of token at YVCurvestETH, assuming a price of 0 for it...
Couldn't find the price of token at ARNXM, assuming a price of 0 for it...
Couldn't find the price of token at MLN, assuming a price of 0 for it...
Couldn't find the price of token at VSP, assuming a price of 0 for it...
Couldn't find the price of token at G, assuming a price of 0 for it...
Couldn't find the price of token at Y3CRV, assuming a price of 0 for it...
Couldn't find the price of token at DUSD, assuming a price of 0 for it...
Couldn't find the price of token at EURS, assuming a price of 0 for it...
Couldn't find the price of token at GUSD, assuming a price of 0 for it...
Couldn't find the price of token at CUSDT, assuming a price of 0 for it...
Couldn't find the price of token at CUSDC, assuming a price of 0 for it...``

**Comment by 0xngmi** [comment]
Thanks, will take a look at it now. The TVL adapter is on projects/creamswap/index.js, is that correct?

**Comment by jeremyHD** [comment]
Well... we are no longer maintaining creamswap. It should be deprecated
I thought we are discussing projects/cream.js here

**Comment by 0xngmi** [comment]
Looking at the code it seems like @petrovska-petro used creamswap/index.js to write an adapter that uses the SDK (instead of just an adapter that returns the TVL as a number, which is what cream.js does) but I'm not sure, so was asking for it. In other words, we are discussing cream but it seems he just chose to reuse another name temporarily.

**Comment by petrovska-petro** [comment]
> Thanks, will take a look at it now. The TVL adapter is on projects/creamswap/index.js, is that correct?

Yes, that is correct. So far, only for Ethereum chain tvl is done.

**Comment by petrovska-petro** [comment]
> Well... we are no longer maintaining creamswap. It should be deprecated
> I thought we are discussing projects/cream.js here

I think that @0xngmi requested also this piece of tvl broken down for each chain that why I took over that file, which I spotted left behind data accuracy wise.

But feel free to point out which is the right file as I am quite confused.

**Pull request merged**